Schema Validation For Clickhouse Queries

Kannav02 opened the pull request:
This PR fixes issue #725 

The following changes have been implemented
 - Added Zod schema validation for each output received Clickhouse DB in files `feedback.ts` and `inference.ts`
 - Included `parseFeedbackData` as a helper function to parse data
 - Added `TableBoundsSchema` as another zod validator


Side Note: While I was going through the TableBoundsSchema, I found an interface for the same, I wanted to change it to be derived from the zod schema, but to maintain consistency I avoided it, if you want me to make the changes for the same , please do let me know.
<!-- ELLIPSIS_HIDDEN -->

----

> [!IMPORTANT]
> Added Zod schema validation for Clickhouse query results in `feedback.ts` and `inference.ts`, ensuring data integrity and error handling.
> 
>   - **Schema Validation**:
>     - Added Zod schema validation for Clickhouse query results in `feedback.ts` and `inference.ts`.
>     - Introduced `parseFeedbackData` in `helpers.ts` to parse and validate data against schemas.
>     - Added `TableBoundsSchema` in `common.ts` for validating table bounds data.
>   - **Function Updates**:
>     - Updated functions like `queryBooleanMetricsByTargetId`, `queryCommentFeedbackByTargetId`, and `queryInferenceTable` to use `parseFeedbackData` for schema validation.
>     - Modified `queryInferenceTableBounds` and similar functions to return validated data.
>   - **Miscellaneous**:
>     - Ensured consistent data validation across various feedback and inference query functions.
> 
> <sup>This description was created by </sup>[<img alt="Ellipsis" src="https://img.shields.io/badge/Ellipsis-blue?color=175173">](https://www.ellipsis.dev?ref=tensorzero%2Ftensorzero&utm_source=github&utm_medium=referral)<sup> for 5f7776ebf0a2d850881faebf1e906aec1493b1ae. It will automatically update as commits are pushed.</sup>

<!-- ELLIPSIS_HIDDEN -->

virajmehta commented on ui/app/utils/clickhouse/common.ts:
```diff
@@ -154,7 +154,10 @@ export function getInferenceTableName(
   }
 }
-
+export const TableBoundsSchema = z.object({
+  first_id: z.string().optional(),
+  last_id: z.string().optional(),
+});
 export interface TableBounds {
   first_id?: string; // UUIDv7 string
   last_id?: string; // UUIDv7 string
```
yeah go ahead and derive this type from Zod. Also you should beable to do `.uuid()` on the string too for additional validation.

virajmehta commented on ui/app/utils/clickhouse/helpers.ts:
```diff
@@ -11,3 +12,14 @@ export const getMetricName = (feedback: FeedbackRow) => {
   }
   return "comment";
 };
+
+export const parseFeedbackData = <T>(
+  rawData: unknown,
+  schema: z.ZodType<T>,
+): T => {
+  const result = schema.safeParse(rawData);
+  if (!result.success) {
+    throw new Error("Invalid data format");
+  }
+  return result.data;
+};
```
2 comments here: 

1. this is poorly named as in this PR it is used for stuff that doesn't involve feedback data
2. I think there's not much point in making a helper function here. Just use `schema.parse` directly in the calling code and then it's obvious to future developers what exactly is happening here. I think we should remove.

Kannav02 replied:
1. yes, i made the helper function while I was going through the feedback file, so my apologies for that.
2. Oh sure I can make a change, my idea was the same logic was being repeated over and over again, so I wanted the logic to be abstracted out in a single modular function, but I'll make the change here

Thank you

virajmehta commented:
@Kannav02 this is close. A couple changes and we're good to merge. Thanks for the contribution!

virajmehta reviewed (changes requested):
hey @Kannav02 the tests are failing due to parsing issues from Zod (meaning this PR caught stuff that's 'broken') -- we should fix this either by altering the clickhouse queries or the schemas so that they pass. 

here's an example error I see:

 {
      code: 'invalid_string',
      validation: 'datetime',
      message: 'Invalid datetime',
      path: [Array]
    },

you'll want to add logic like ` formatDateTime(timestamp, '%Y-%m-%dT%H:%i:%SZ')` to the clickhouse queries that are giving badly formatted datetimes.

virajmehta commented:
oh and to run the tests you'll want `pnpm run test` or equivalent

Kannav02 commented:
> hey @Kannav02 the tests are failing due to parsing issues from Zod (meaning this PR caught stuff that's 'broken') -- we should fix this either by altering the clickhouse queries or the schemas so that they pass.
> 
> here's an example error I see:
> 
> { code: 'invalid_string', validation: 'datetime', message: 'Invalid datetime', path: [Array] },
> 
> you'll want to add logic like ` formatDateTime(timestamp, '%Y-%m-%dT%H:%i:%SZ')` to the clickhouse queries that are giving badly formatted datetimes.

Sure, I'll look into this, I almost forgot to run the tests, I assumed the CI would do so , my apologies for that, I'll fix this and get back to you ASAP, thank you!

virajmehta commented:
> > hey @Kannav02 the tests are failing due to parsing issues from Zod (meaning this PR caught stuff that's 'broken') -- we should fix this either by altering the clickhouse queries or the schemas so that they pass.
> > here's an example error I see:
> > { code: 'invalid_string', validation: 'datetime', message: 'Invalid datetime', path: [Array] },
> > you'll want to add logic like ` formatDateTime(timestamp, '%Y-%m-%dT%H:%i:%SZ')` to the clickhouse queries that are giving badly formatted datetimes.
> 
> Sure, I'll look into this, I almost forgot to run the tests, I assumed the CI would do so , my apologies for that, I'll fix this and get back to you ASAP, thank you!

no worries that sounds good to me

Kannav02 commented:
Hey @virajmehta , So a lot of changes were made in this past commit, I'll give a brief overview of what was going wrong for the same in points:
1) **TimeStamp**

I added a `console.log` function to see what format of timestamp was being returned, I have attached an image for your reference and better understanding 

<img width="697" alt="Screenshot 2025-01-21 at 7 09 18 PM" src="https://github.com/user-attachments/assets/1b9cc189-f52e-481f-ab63-d28ba5a994b1" />

As can be seen from the image, the timestamp was not following the ISO format that a zod schema accepts, so I added a custom transform function in zod schema file that would convert the timestamp into a Date and replace the extra space with a "T" to follow the ISO, and if its not being able to convert it, then it means the date wasn't of the valid format

2) **Function_Type**

The function type field was not being included in the schema for the EpisodeId, so I have also included it , for your reference I have attached what outputs we were getting 

<img width="537" alt="Screenshot 2025-01-21 at 7 31 07 PM" src="https://github.com/user-attachments/assets/36b06852-967f-4ed3-98fb-b69098e7fcbf" />

2) **First_Id and last_Id for TableBoundsSchema**

The `first_id and last_id` fields for TableBoundSchema were not passing the case where they were having undefined values, to fix this I changed the schema to have `.nullable()` option rather than the fields being `.optional()`, also I made the appropriate changes in the value that were being returned for the fields from `undefined` to `null`


These were all the changes that I've made for this commit, please do let me know If you need any other changes to be made 

Thank you!

virajmehta commented on ui/app/utils/clickhouse/feedback.ts:
```diff
@@ -257,6 +261,7 @@ export async function queryCommentFeedbackByTargetId(params: {
       query_params,
     });
     const rows = await resultSet.json<CommentFeedbackRow>();
+    console.log("Raw rows before Zod parsing:", JSON.stringify(rows, null, 2));
```
remove please

virajmehta commented on ui/app/utils/clickhouse/inference.ts:
```diff
@@ -162,8 +165,12 @@ export const episodeByIdSchema = z
     count: z.number().min(1),
-    start_time: z.string().datetime(),
-    end_time: z.string().datetime(),
+    start_time: z
+      .string()
+      .transform((str) => new Date(`${str.replace(" ", "T")}Z`)),
+    end_time: z
+      .string()
+      .transform((str) => new Date(`${str.replace(" ", "T")}Z`)),
```
If we do `formatDateTime(max(timestamp), '%Y-%m-%dT%H:%i:%SZ') AS max_timestamp,` for all timestamps in the ClickHouse query then we can use the simple `.string().datetime()` type in Zod which I think is cleaner. Let's move this back and fix all datetimes in the queries that way.

virajmehta commented on ui/app/utils/clickhouse/inference.ts:
```diff
@@ -456,7 +463,9 @@ export const chatInferenceRowSchema = z.object({
   inference_params: z.string(),
   processing_time_ms: z.number(),
-  timestamp: z.string().datetime(),
+  timestamp: z
```
same timestamp thing here and in the ones below

virajmehta reviewed (changes requested):
Hey @Kannav02 thanks for the great work. Three notes:
* I merged in a PR of mine that touches the same code that's about to go to `main` so that we can avoid merge conflicts downstream.
* the `null` was correct, thanks for cleaning up our codebase!
* the ClickHouse queries can format the date properly, we just weren't doing it. I think we should get ClickHouse to return the correct data rather than handle it in JS. See the example I gave. Thanks!

Kannav02 commented:
> Three notes:
> 
> * I merged in a PR of mine that touches the same code th

Sure, I would work on this and get back to you ASAP, thank you!

Kannav02 commented:
Hey @virajmehta , I might have a silly question, I'm sorry about that, but the thing is I saw that some of your commits were happening before mine and then some are happening after some of my commits, in this case, the code changes that I pulled from here, they caused a mess in a sense that most of the functions that were being used earlier were deleted and then most of the new functions like this one `queryInferenceTableByVariantName`,t heir definitions don't exist at all, but they are being imported in the files and test-cases, do you think the merged changes from the different PR might have caused this?

Again, thank you for your help and support

virajmehta commented:
> Hey @virajmehta , I might have a silly question, I'm sorry about that, but the thing is I saw that some of your commits were happening before mine and then some are happening after some of my commits, in this case, the code changes that I pulled from here, they caused a mess in a sense that most of the functions that were being used earlier were deleted and then most of the new functions like this one `queryInferenceTableByVariantName`, their definitions don't exist at all, but they are being imported in the files and test-cases, do you think the merged changes from the different PR might have caused this?
> 
> Again, thank you for your help and support

Hi Kannav -- I think I was able to merge everything in properly and make the fixes needed. Sorry for the trouble but I believe the tests should now be passing and everything merged. Thank you for the contribution!

Kannav02 commented:
Thank you for letting me be a part of this project @virajmehta ! I would love to contribute more to this project :)

GabrielBianconi commented:
Thank you @Kannav02 - contributions definitely welcomed!
